module lop.error;

members {
    Error
}

// TODO: since we have the notion of an error being a result, we should
// rename this to be Description (of a failure/error) or somesuch

def Error class(expected: String, actual: String) => {
    members {
        expected fun() => expected,
        actual fun() => actual,
        equals,
        represent
    }
    
    def equals fun(other: Error) =>
        and(expected.equals(other.expected()), actual.equals(other.actual()));
        
    def represent fun() =>
        "expected ".concat(expected).concat(" but got ").concat(actual);
};
